modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryHandler.java

```diff
@@ -1431,27 +1424,42 @@ private void sendBackupAcknowledge(final IgniteBiTuple<Map<Integer, Long>, Set<A
                         routineId,
                         t.get1());
 
+                    Set<UUID> sentToNodes = t.get2().size() > 1 ? new HashSet<>() : null;
+
                     for (AffinityTopologyVersion topVer : t.get2()) {
-                        for (ClusterNode node : ctx.discovery().cacheGroupAffinityNodes(cctx.groupId(), topVer)) {
-                            if (!node.isLocal() && cctx.discovery().alive(nodeId)) {
-                                try {
-                                    cctx.io().send(node, msg, GridIoPolicy.SYSTEM_POOL);
-                                }
-                                catch (ClusterTopologyCheckedException ignored) {
-                                    IgniteLogger log = ctx.log(CU.CONTINUOUS_QRY_LOG_CATEGORY);
+                        if (topVer == null)
+                            topVer = AffinityTopologyVersion.NONE;
 
-                                    if (log.isDebugEnabled())
-                                        log.debug("Failed to send acknowledge message, node left " +
-                                            "[msg=" + msg + ", node=" + node + ']');
-                                }
-                                catch (IgniteCheckedException e) {
-                                    IgniteLogger log = ctx.log(CU.CONTINUOUS_QRY_LOG_CATEGORY);
+                        try {
+                            for (ClusterNode node : ctx.discovery().cacheGroupAffinityNodes(cctx.groupId(), topVer)) {
+                                if (!node.isLocal() && cctx.discovery().alive(nodeId)) {
```

Comment:
do we need additional check for nodeId nullability here ?

Reply:
no `nodeId` can't be null.